### PR TITLE
feat: Custom cape and elytra cosmetics

### PR DESCRIPTION
**Pull request by P0keDev**
Essentially ports legacy's cape and elytra handling. Custom `RenderLayer` objects are attached to the vanilla player render engine to handle our versions of the respective cosmetics, and any capes or elytras being rendered through the regular layer will be disabled if ours is active.

I had to change the way that `PlayerModel` loaded players, since the original `PlayerJoinedWorldEvent` was implemented in such a way that it never triggered on the user's own account.

image

### CODE REVIEW

File: common/src/main/java/com/wynntils/features/user/players/WynntilsCosmeticsFeature.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                float netHeadYaw,
+                float headPitch) {
+            if (!parent.shouldRenderCosmetic(player, Type.CAPE)) return;
+
+            ResourceLocation texture = parent.getTexture(player);
+            if (texture == null) return;
+
+            poseStack.pushPose();
+            poseStack.translate(0.0f, 0.0f, 0.125f);
+            double d = Mth.lerp(partialTick, player.xCloakO, player.xCloak)
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/players/WynntilsCosmeticsFeature.java]: Is this math copied from MC source?

File: common/src/main/java/com/wynntils/mc/mixin/ClientLevelMixin.java [hunk 2]
```diff
@@ -6,13 +6,19 @@
...
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 @Mixin(ClientLevel.class)
 public abstract class ClientLevelMixin {
+    @Inject(method = "addPlayer(ILnet/minecraft/client/player/AbstractClientPlayer;)V", at = @At("HEAD"))
+    private void addPlayer(int id, AbstractClientPlayer player, CallbackInfo ci) {
+        EventFactory.onPlayerJoinedWorld(player);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/mixin/ClientLevelMixin.java:19]: To be sure, this is the change so we load our own user's info too?

**Review by kristofbolyai — DISMISSED** [review]
Looks good overall. Do you want to review this or are we good to merge? @magicus

File: common/src/main/java/com/wynntils/mc/mixin/ClientLevelMixin.java (hunk 2 above)
Comment (P0keDev) [reply common/src/main/java/com/wynntils/mc/mixin/ClientLevelMixin.java:19]: It is, yes. The method we used to mix into delegated to this one, actually - but it was the packet listener class, so it was only handling other players

File: common/src/main/java/com/wynntils/features/user/players/WynntilsCosmeticsFeature.java (hunk 1 above)
Comment (P0keDev) [reply common/src/main/java/com/wynntils/features/user/players/WynntilsCosmeticsFeature.java]: yupp

i can try to clean it up if you want? this code is for sure a huge mess

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/players/WynntilsCosmeticsFeature.java]: I think this is different from other cases, where we want to fully preserve mc code. So yes, cleanup please.

**Comment by magicus** [comment]
Let me have a look later tonight.

**Review by magicus — DISMISSED** [review]
I think it looks good. You have a trivial-ish merge conflict from my changes in PlayerModel incoming. I can't read and understand the rendering code, but if bolyai wants cleaner gibberish, then you should probably do that :-)

**Review by magicus — APPROVED** [review]

**Pull request merged**